### PR TITLE
Issue 3385: Extract stripe module

### PR DESCRIPTION
**Pull request by santail**
## Context behind the change

https://github.com/humanprotocol/human-protocol/issues/3385

It was decided to extract payment providers specific logic to separate module as abstraction and Stripe service implementing it. It would be easily to replace Stripe with any other service which should implement respective abstract class. 

Now payment service operates with payment provider abstraction. 

hmt.users::stripe_customer_id table column was renamed to payment_provider_id

## How has this been tested?

Services are covered with unit tests. Some tests were extended, some were added or adjusted. 
Verified on local environment by walking through new job creation flow in job-creator. 

## Release plan
Update env variable names:
```
PAYMENT_PROVIDER_SECRET_KEY
PAYMENT_PROVIDER_APP_NAME
PAYMENT_PROVIDER_APP_VERSION
PAYMENT_PROVIDER_APP_INFO_URL
PAYMENT_PROVIDER_API_VERSION
```

### CODE REVIEW

**Review by flopez7 — CHANGES_REQUESTED** [review]
After discussing this with @portuu3, we agreed we should change the approach from creating a separate Stripe module to using an abstract class for the payment provider instead. This way, we can keep general logic in PaymentService and have the Stripe-specific implementation that extend the abstract class. We’re already using this pattern in other parts of the project, like [email service](https://github.com/humanprotocol/human-protocol/tree/develop/packages/apps/reputation-oracle/server/src/modules/email) with sendgrid in the reputation oracle. This way is easier to maintain or switch providers in the future.

For this, we would define an abstract base class for payment providers that contains any shared logic (if needed) and the abstract methods for the provider-specific actions (like creating a customer, creating an invoice or processing a payment). The Stripe implementation would simply extend this abstract class and provide the Stripe-specific logic. This keeps the PaymentService decoupled from any single provider and makes it straightforward to add or switch payment providers later, since all common code is in one place and each provider only needs to implement the unique parts.

**Review by flopez7 — CHANGES_REQUESTED** [review]
- Move provider implementation into a dedicated folder inside payment, e.g. `modules/payment/providers/stripe.provider.ts`
- Place the abstract class in `modules/payment/providers/payment-provider.abstract.ts`
- Generalize provider method names in the abstract class so they are not Stripe-specific. For example, avoid names like createSetupIntent or handlePaymentIntent unless they are generic enough for all providers.

**Review by flopez7 — CHANGES_REQUESTED** [review]
- Move provider implementation into a dedicated folder inside payment, e.g. `modules/payment/providers/stripe.provider.ts`
- Place the abstract class in `modules/payment/providers/payment-provider.abstract.ts`
- Generalize provider method names in the abstract class so they are not Stripe-specific. For example, avoid names like createSetupIntent or handlePaymentIntent unless they are generic enough for all providers.

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+    return {
+      id: invoice.id,
+      payment_id: invoice.payment_intent as string,
+      status: invoice.status?.toString(),
+      amount_due: invoice.amount_due,
+      currency: invoice.currency,
+    } as Invoice;
+  }
+
+  async createPayment(
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:80]: Rename `createPayment` method to something more descriptive, like `assignPaymentMethod`. The current name is a bit misleading, since this method is not actually creating a payment but rather assigning a payment method and confirming the payment intent

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+            postal_code: customer.address.postal_code ?? undefined,
+          }
+        : undefined,
+      default_payment_method: customer.invoice_settings
+        ? (customer.invoice_settings.default_payment_method as string)
+        : undefined,
+    };
+  }
+
+  async listCustomerTaxIds(customerId: string): Promise<TaxId[]> {
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:358]: Should be private

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    const taxIds = await this.stripe.customers.listTaxIds(customerId);
+
+    return taxIds.data.map((taxId) => ({
+      id: taxId.id,
+      type: taxId.type as VatType,
+      value: taxId.value,
+    }));
+  }
+
+  async createTaxId(
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:368]: should be private

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+      value,
+    });
+    return {
+      id: taxId.id,
+      type: taxId.type as VatType,
+      value: taxId.value,
+    };
+  }
+
+  async deleteTaxId(customerId: string, taxIdId: string): Promise<void> {
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:384]: Should be private

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+    it('should handle missing client secret', async () => {
+      stripeMock.setupIntents.create = jest.fn().mockResolvedValue({});
+
+      await expect(service.setupCard('cus_123')).rejects.toThrow(
+        new ServerError(ErrorPayment.ClientSecretDoesNotExist),
+      );
+    });
+  });
+
+  describe('handlePaymentIntent', () => {
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts:160]:
```suggestion
  describe('assignPaymentMethod', () => {
```

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+        .mockRejectedValue(new Error('Stripe error'));
+
+      await expect(service.createCustomer('test@example.com')).rejects.toThrow(
+        new ServerError(ErrorPayment.CustomerNotCreated),
+      );
+      expect(loggerSpy).toHaveBeenCalled();
+    });
+  });
+
+  describe('createSetupIntent', () => {
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts:116]:
```suggestion
  describe('setupCard', () => {
```

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+import { NotFoundError, ServerError } from '../../../../common/errors';
+import { ErrorPayment } from '../../../../common/constants/errors';
+import {
+  PaymentCurrency,
+  PaymentStatus,
+  VatType,
+} from '../../../../common/enums/payment';
+import { PaymentProvider } from '../payment-provider.abstract';
+
+jest.mock('stripe');
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts:15]: This should go at the beginning of the file

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,726 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.spec.ts:1]: Great work with tests!
Since we're improving all our tests (you can check `/job/job.service.spec.ts`), I suggest the following:

- Add fixtures (in fixtures.ts) for common test data if you find yourself repeating mock objects or DTOs, e.g. setupIntent, paymentIntent, etc.
- Use faker to generate random and realistic test data. This helps cover more edge cases and makes the tests more robust.
- Please add tests for the following public methods, as they are not currently covered:
  - retrievePaymentIntent
  - getDefaultPaymentMethod
  - listPaymentMethods
  - retrieveCardSetup
- Also make sure you test all scenarios

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 9]
```diff
@@ -1,6 +1,60 @@
...
+  expMonth: number;
+  expYear: number;
+  default: boolean;
+}
+
+export interface BillingAddress {
+  line1?: string;
+  city?: string;
+  country?: string;
+  postal_code?: string;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:22]:
```suggestion
  postalCode?: string;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 10]
```diff
@@ -1,6 +1,60 @@
...
+}
+
+export interface CustomerData {
+  email: string;
+  name?: string;
+  address?: BillingAddress;
+  default_payment_method?: string;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:29]:
```suggestion
  defaultPaymentMethod?: string;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 11]
```diff
@@ -1,6 +1,60 @@
...
+
+export interface TaxId {
+  id: string;
+  type: VatType;
+  value: string;
+}
+
+export interface Invoice {
+  id: string;
+  payment_id: string | null;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:40]:
```suggestion
  paymentId: string | null;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 12]
```diff
@@ -1,6 +1,60 @@
...
+  status?: string;
+  amount_due: number;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:42]:
```suggestion
  amountDue: number;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 13]
```diff
@@ -1,6 +1,60 @@
...
+  currency: string;
+}
+
+export interface CardSetup {
+  customer_id: string;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:47]:
```suggestion
  customerId: string;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 14]
```diff
@@ -1,6 +1,60 @@
...
+  payment_method: string;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:48]:
```suggestion
  paymentMethod: string;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 15]
```diff
@@ -1,6 +1,60 @@
...
+}
+
+export interface PaymentData {
+  customer: string;
+  id: string;
+  client_secret: string | null;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:54]:
```suggestion
  clientSecret: string | null;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 16]
```diff
@@ -1,6 +1,60 @@
...
+  status: string;
+  amount: number;
+  amount_received: number;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:57]:
```suggestion
  amountReceived: number;
```

File: packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts [hunk 17]
```diff
@@ -1,6 +1,60 @@
...
+  currency: string;
+  latest_charge: string;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/payment/payment.interface.ts:59]:
```suggestion
  latestCharge: string;
```

**Review by flopez7 — CHANGES_REQUESTED** [review]
We should use camelCase for all interface property names to follow TypeScript and JavaScript conventions. Please update all internal references

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts (hunk 1 above)
Comment (flopez7) [reply packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:80]: The paymentIntent is created by Stripe automatically when the invoice is finalized. This method just attach the right payment method and set it as offSession

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts [hunk 18]
```diff
@@ -0,0 +1,420 @@
...
+            city,
+            country,
+            postal_code: postalCode,
+          },
+        };
+
+    const customer = (await this.stripe.customers.update(
+      customerId,
+      updatePayload,
+    )) as Stripe.Customer;
```
[the remark this answers is not in the document]
Comment (santail) [reply packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:328]: This casting is required as stripe.customers.update and stripe.customers.retrieve return ambiguous result type.

File: .gitignore [hunk 19]
```diff
@@ -48,4 +48,5 @@ dist
 hardhat-dependency-compiler
 
 # cache
-cache
\ No newline at end of file
+cache
+*.iml
```
[the remark this answers is not in the document]
Comment (santail) [reply .gitignore:52]: That is configuration file for IntelliJ Idea projects. Should be ignored for everyone I suppose. 
Not the best place for that change in current PR. But still should be done.

File: packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (flopez7) [reply packages/apps/job-launcher/server/src/modules/payment/providers/stripe/stripe.service.ts:332]: We don't actually need the casting here, since we only use the returned object to extract specific properties and build our own CustomerData object.
Stripe.Response<Stripe.Customer> already includes all customer properties, so we can safely access them without casting

**Review by flopez7 — CHANGES_REQUESTED** [review]

**Review by portuu3 — CHANGES_REQUESTED** [review]
In order to fully encapsulate it:
- Move this enum inside sendgrid module https://github.com/humanprotocol/human-protocol/blob/44d7635faee9f9aa87e0a316ebc5cf833d39b973/packages/apps/job-launcher/server/src/common/enums/payment.ts#L39
- Rename StripeConfigService to PaymentProviderConfigService (also rename its properties and envs used)

**Review by flopez7 — APPROVED** [review]
